trame_server/core.py

```diff
@@ -559,7 +559,14 @@ def start(
         if exec_mode == "main":
             self._running_stage = 0
             if self.controller.on_server_exited.exists():
-                self.controller.on_server_exited(**self.state.to_dict())
+                loop = asyncio.get_event_loop()
+                for exit_task in self.controller.on_server_exited(
+                    **self.state.to_dict()
+                ):
+                    if isinstance(exit_task, asyncio.Future):
+                        loop.run_until_complete(exit_task)
+                    elif callable(exit_task):
+                        exit_task()
```

Comment:
So it might be better to do the following

```python
if inspect.isawaitable(exit_task):
   loop.run_until_complete(exit_task)
elif callable(exit_task):
   result = exit_task()
   if inspect.isawaitable(result):
     loop.run_until_complete(result)
```

Not sure why you could have a callable in the result, but why not...

Reply:
This works. I was trying various types of exit callbacks and wrote one that returned a callable wrapped in a future. A little strange, but yes, why not.